File: java/com/google/copybara/util/CommandRunner.java
/*
 * Copyright (C) 2016 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.copybara.util;

import static java.nio.charset.StandardCharsets.UTF_8;

import com.google.common.base.Preconditions;
import com.google.common.base.Splitter;
import com.google.common.base.Stopwatch;
import com.google.common.collect.ImmutableList;
import com.google.common.flogger.FluentLogger;
import com.google.errorprone.annotations.CheckReturnValue;
import com.google.copybara.shell.AbnormalTerminationException;
import com.google.copybara.shell.BadExitStatusException;
import com.google.copybara.shell.Command;
import com.google.copybara.shell.CommandException;
import com.google.copybara.shell.Killable;
import com.google.copybara.shell.KillableObserver;
import com.google.copybara.shell.ShellUtils;
import com.google.copybara.shell.TerminationStatus;
import com.google.copybara.shell.TimeoutKillableObserver;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.OutputStream;
import java.time.Duration;
import java.time.LocalTime;
import java.time.format.DateTimeFormatter;
import java.util.Arrays;
import java.util.Optional;
import java.util.logging.Level;

/**
 * Allows running a {@link Command} with easier stderr/stdout and logging management.
 */
public final class CommandRunner {

  private static final FluentLogger logger = FluentLogger.forEnclosingClass();

  /**
   * No input for the command.
   */
  public static final byte[] NO_INPUT = new byte[]{};
  // By default we kill the command after 15 minutes.
  public static final Duration DEFAULT_TIMEOUT = Duration.ofMinutes(15);
  public static final int MAX_COMMAND_LENGTH = 40000;

  private final Command cmd;
  private final boolean verbose;
  private final byte[] input;
  private final int maxOutLogLines;
  private final Duration timeout;
  private final ImmutableList<KillableObserver> additionalObservers;
  private final Optional<OutputStream> asyncStdoutStream;
  private final Optional<OutputStream> asyncErrStream;
  private final Optional<CommandExecutor> executor;


  private CommandRunner(Command cmd, boolean verbose, byte[] input, int maxOutLogLines,
      Duration timeout,
      ImmutableList<KillableObserver> additionalObservers,
      Optional<OutputStream> stdoutStream,
      Optional<OutputStream> errStream,
      Optional<CommandExecutor> executor) {
    this.cmd = Preconditions.checkNotNull(cmd);
    this.verbose = verbose;
    this.input = Preconditions.checkNotNull(input);
    this.maxOutLogLines = maxOutLogLines;
    this.timeout = Preconditions.checkNotNull(timeout);
    this.additionalObservers = Preconditions.checkNotNull(additionalObservers);
    this.asyncStdoutStream = Preconditions.checkNotNull(stdoutStream);
    this.asyncErrStream = Preconditions.checkNotNull(errStream);
    this.executor = Preconditions.checkNotNull(executor);
  }

  public CommandRunner(Command cmd) {
    this(cmd, false, NO_INPUT, -1, DEFAULT_TIMEOUT, ImmutableList.of(),
        Optional.empty(), Optional.empty(), Optional.empty());
  }

  public CommandRunner(Command cmd, Duration timeout) {
    this(cmd, false, NO_INPUT, -1, timeout, ImmutableList.of(),
        Optional.empty(), Optional.empty(), Optional.empty());
  }

  /**
   * Sets the verbose level for the command execution.
   */
  @CheckReturnValue
  public CommandRunner withVerbose(boolean verbose) {
    return new CommandRunner(
        this.cmd, verbose, this.input, this.maxOutLogLines, timeout, additionalObservers,
        asyncStdoutStream, asyncErrStream, executor);
  }

  /**
   * Sets the input for the command execution.
   */
  @CheckReturnValue
  public CommandRunner withInput(byte[] input) {
    return new CommandRunner(
        this.cmd, this.verbose, input, this.maxOutLogLines, timeout, additionalObservers,
        asyncStdoutStream, asyncErrStream, executor);
  }

  /**
   * Sets the maximum number of output lines logged per stream
   */
  @CheckReturnValue
  public CommandRunner withMaxStdOutLogLines(int lines) {
    return new CommandRunner(
        this.cmd, this.verbose, this.input, lines, timeout, additionalObservers,
        asyncStdoutStream, asyncErrStream, executor);
  }

  /**
   * Adds another observer to the call executions
   */
  @CheckReturnValue
  public CommandRunner withObserver(KillableObserver observer) {
    return new CommandRunner(
        this.cmd, this.verbose, this.input, maxOutLogLines, timeout,
        ImmutableList.<KillableObserver>builder().addAll(additionalObservers).add(observer).build(),
        asyncStdoutStream, asyncErrStream, executor);
  }

  /**
   * Sets a stream to redirect stdOut output to
   */
  @CheckReturnValue
  public CommandRunner withStdOutStream(OutputStream stream) {
    return new CommandRunner(
        this.cmd, this.verbose, this.input, maxOutLogLines, timeout,
        additionalObservers,
        Optional.ofNullable(stream),
        asyncErrStream,
        executor);
  }

  /**
   * Sets a stream to redirect stdErr output to
   */
  @CheckReturnValue
  public CommandRunner withStdErrStream(OutputStream stream) {
    return new CommandRunner(
        this.cmd, this.verbose, this.input, maxOutLogLines, timeout,
        additionalObservers,
        asyncStdoutStream,
        Optional.ofNullable(stream),
        executor);
  }

  /**
   * Sets a stream to redirect stdErr output to
   */
  @CheckReturnValue
  public CommandRunner withCommandExecutor(CommandExecutor runner) {
    return new CommandRunner(
        this.cmd, this.verbose, this.input, maxOutLogLines, timeout,
        additionalObservers,
        asyncStdoutStream,
        asyncErrStream,
        Optional.of(runner));
  }

  /**
   * Executes a {@link Command} with the given input and writes to the console and the log depending
   * on the exit code of the command and the verbose flag.
   */
  public CommandOutputWithStatus execute() throws CommandException {
    Stopwatch stopwatch = Stopwatch.createStarted();
    String startMsg = ShellUtils.prettyPrintArgv(Arrays.asList(cmd.getCommandLineElements()));
    startMsg = startMsg.length() > MAX_COMMAND_LENGTH
        ? startMsg.substring(0, MAX_COMMAND_LENGTH) + "..." : startMsg;
    String validStartMsg = "Executing [" + startMsg + "]";
    logger.atInfo().log("%s", validStartMsg);
    if (verbose) {
      System.err.println(validStartMsg);
    }
    TerminationStatus exitStatus = null;
    CombinedKillableObserver cmdMonitor =
        new CombinedKillableObserver(timeout, additionalObservers.toArray(new KillableObserver[0]));
    ByteArrayOutputStream stdoutCollector = new ByteArrayOutputStream();
    ByteArrayOutputStream stderrCollector = new ByteArrayOutputStream();
    try {
      if (asyncStdoutStream.isPresent()) {
        stdoutCollector.write("stdOut redirected to external observer.".getBytes(UTF_8));
      }
      if (asyncErrStream.isPresent()) {
        stderrCollector.write("stdErr redirected to external observer.".getBytes(UTF_8));
      }
    } catch (IOException e) {
      logger.atSevere().withCause(e).log("Error writing output.");
    }
    OutputStream stdoutStream = commandOutputStream(asyncStdoutStream.orElse(stdoutCollector));
    OutputStream stderrStream = commandOutputStream(asyncErrStream.orElse(stderrCollector));

    try {
      CommandExecutor runner = executor.orElse(new DefaultExecutor());
      TerminationStatus status =
          runner.getCommandOutputWithStatus(cmd, input, cmdMonitor, stdoutStream, stderrStream);
      exitStatus = status;
      return new CommandOutputWithStatus(
          status,
          stdoutCollector.toByteArray(),
          stderrCollector.toByteArray());
    } catch (BadExitStatusException e) {
      exitStatus = e.getResult().getTerminationStatus();
      maybeTreatTimeout(stdoutCollector, stderrCollector, cmdMonitor, e);
      throw new BadExitStatusWithOutputException(e.getCommand(), e.getResult(), e.getMessage(),
          stdoutCollector.toByteArray(),
          stderrCollector.toByteArray());
    } catch (AbnormalTerminationException e) {
      maybeTreatTimeout(stdoutCollector, stderrCollector, cmdMonitor, e);
      throw e;
    } finally {
      String commandName = cmd.getCommandLineElements()[0];

      if (maxOutLogLines != 0) {
        logOutput(
            Level.INFO,
            String.format("'%s' STDOUT: ", commandName),
            stdoutCollector,
            maxOutLogLines);
        logOutput(
            Level.INFO,
            String.format("'%s' STDERR: ", commandName),
            stderrCollector,
            maxOutLogLines);
      }

      String finishMsg;
      if (cmdMonitor.hasTimedOut()) {
        finishMsg = String.format(
            "Command '%s' was killed after timeout. Execution time %s. %s",
            commandName, formatDuration(stopwatch.elapsed()),
            exitStatus != null ? exitStatus.toString() : "(No exit status)");
        logger.atSevere().log("%s", finishMsg);
      } else {
        finishMsg = String.format(
            "Command '%s' finished in %s. %s",
            commandName, formatDuration(stopwatch.elapsed()),
            exitStatus != null ? exitStatus.toString() : "(No exit status)");
        logger.atInfo().log("%s", finishMsg);
      }
      if (verbose) {
        System.err.println(finishMsg);
      }
    }
  }

  /**
   * Format a duration to a human-readable string. This assumes that the duration is less than
   * 24 hours, which should always be true for a command (usually takes from a few ms to a few
   * minutes).
   */
  private String formatDuration(Duration duration) {
    return LocalTime.MIDNIGHT.plus(duration).format(DateTimeFormatter.ofPattern("mm:ss.SSS"));
  }

private static class DefaultExecutor implements CommandExecutor {

  @Override
  public TerminationStatus getCommandOutputWithStatus(Command cmd, byte[] input,
      KillableObserver cmdMonitor, OutputStream stdoutStream, OutputStream stderrStream)
      throws CommandException {
    return cmd.execute(input, cmdMonitor, stdoutStream, stderrStream, true)
        .getTerminationStatus();
  }
}

  private void maybeTreatTimeout(ByteArrayOutputStream stdoutCollector,
      ByteArrayOutputStream stderrCollector, CombinedKillableObserver cmdMonitor,
      AbnormalTerminationException e) throws CommandTimeoutException {
    if (!cmdMonitor.hasTimedOut()) {
      return;
    }
    String msg = String.format(
        "Command '%s' killed by Copybara after timeout (%ds)."
            + " If this fails during a fetch use --fetch-timeout flag.\n"
            + "Exit info: %s",
        cmd.getCommandLineElements()[0],
        timeout.getSeconds(),
        e.getResult().getTerminationStatus());
    throw new CommandTimeoutException(e.getCommand(), e.getResult(), msg,
        stdoutCollector.toByteArray(),
        stderrCollector.toByteArray(),
        timeout);
  }

  /**
   * Creates the necessary OutputStream to be passed to the {@link Command#execute()}.
   */
  private OutputStream commandOutputStream(OutputStream outputStream) {
    // If verbose we stream to the user console too
    return verbose ? new MultiplexOutputStream(System.err, outputStream) : outputStream;
  }

  /**
   * Log to the appropriate log level the output of the command
   */
  private static void logOutput(
      Level level, String prefix, ByteArrayOutputStream outputBytes, int maxLogLines) {
    String string = asString(outputBytes).trim();
    if (string.isEmpty()) {
      return;
    }
    int lines = 0;
    for (String line : Splitter.on(System.lineSeparator()).split(string)) {
      logger.at(level).log("%s%s", prefix, line);
      lines++;
      if (maxLogLines >= 0 && lines >= maxLogLines) {
        logger.at(level).log("%s... truncated after %d line(s)", prefix, maxLogLines);
        break;
      }
    }
  }

  private static String asString(ByteArrayOutputStream outputBytes) {
    return new String(outputBytes.toByteArray(), UTF_8);
  }

  /**
   * An {@link OutputStream} that can output to two {@code OutputStream}
   */
  private static class MultiplexOutputStream extends OutputStream {

    private final OutputStream s1;
    private final OutputStream s2;

    private MultiplexOutputStream(OutputStream s1, OutputStream s2) {
      this.s1 = s1;
      this.s2 = s2;
    }

    @Override
    public void write(int b) throws IOException {
      s1.write(b);
      s2.write(b);
    }

    @Override
    public void write(byte[] b, int off, int len) throws IOException {
      s1.write(b, off, len);
      s2.write(b, off, len);
    }

    @Override
    public void flush() throws IOException {
      s1.flush();
      s2.flush();
    }

    @Override
    public void close() throws IOException {
      IOException ex = null;
      try {
        s1.close();
      } catch (IOException e) {
        ex = e;
      }
      // We favor the second exception over the first one on close. But this is fine for now.
      s2.close();
      if (ex != null) {
        throw ex;
      }
    }
  }

  /**
   * Multiplex KillableObserver to allow monitoring processes
   */
  private static class CombinedKillableObserver implements KillableObserver {

    private final TimeoutKillableObserver timed;
    private final ImmutableList<KillableObserver> others;

    private CombinedKillableObserver(Duration timeout, KillableObserver... others) {
      this.timed = new TimeoutKillableObserver(timeout);
      this.others = ImmutableList.copyOf(others);
    }

    @Override
    public void startObserving(Killable killable) {
      timed.startObserving(killable);
      for (KillableObserver other : others) {
        other.startObserving(killable);
      }
    }

    @Override
    public void stopObserving(Killable killable) {
      timed.stopObserving(killable);
      for (KillableObserver other : others) {
        other.stopObserving(killable);
      }
    }

    /**
     * Returns true if the observed process was killed by this observer.
     */
    public boolean hasTimedOut() {
      return timed.hasTimedOut();
    }
  }

  /**
   * Interface to insert custom command options not otherwise supported by the type.
   */
  public interface CommandExecutor {

    /**
     * Actual invocation of cmd.Execute. This hook is required as implementations of Command differ
     * between versions and implementations. Use with caution.
     *
     * @param cmd cmd to to be run
     * @param input input
     * @param cmdMonitor Observer for terminating job
     * @param stdoutStream Stream for cmd.execute
     * @param stderrStream Stream for cmd.execute
     */
    TerminationStatus getCommandOutputWithStatus(
        Command cmd,
        byte[] input,
        KillableObserver cmdMonitor,
        OutputStream stdoutStream,
        OutputStream stderrStream)
        throws CommandException;
  }
}

File: src/Copybara.Core/Util/CommandRunner.cs
/*
 * Copyright (C) 2016 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

using System.Diagnostics;
using System.Text;
using Copybara.Common;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Logging.Abstractions;

namespace Copybara.Util;

/// <summary>
/// Allows running a <see cref="Command"/> with easier stderr/stdout and logging management. Port of
/// <c>com.google.copybara.util.CommandRunner</c>. Execution is performed with
/// <c>System.Diagnostics.Process</c> rather than the Bazel shell library; stdout/stderr are captured,
/// a timeout kills the process, and the exit status is reported through
/// <see cref="CommandOutputWithStatus"/> and the command exception hierarchy.
/// </summary>
public sealed class CommandRunner
{
    private static readonly ILogger Logger = NullLogger.Instance;

    /// <summary>No input for the command.</summary>
    public static readonly byte[] NoInput = Array.Empty<byte>();

    /// <summary>By default, we kill the command after 15 minutes.</summary>
    public static readonly TimeSpan DefaultTimeout = TimeSpan.FromMinutes(15);

    public const int MaxCommandLength = 40000;

    public static readonly byte[] AfterLimitSuffix =
        Encoding.UTF8.GetBytes("... (Rest of the output skipped)\n");

    private readonly Command _cmd;
    private readonly bool _verbose;
    private readonly byte[] _input;
    private readonly int _maxOutLogLines;
    private readonly TimeSpan _timeout;
    private readonly Stream? _asyncStdoutStream;
    private readonly Stream? _asyncErrStream;

    private CommandRunner(
        Command cmd,
        bool verbose,
        byte[] input,
        int maxOutLogLines,
        TimeSpan timeout,
        Stream? asyncStdoutStream,
        Stream? asyncErrStream)
    {
        _cmd = Preconditions.CheckNotNull(cmd);
        _verbose = verbose;
        _input = Preconditions.CheckNotNull(input);
        _maxOutLogLines = maxOutLogLines;
        _timeout = timeout;
        _asyncStdoutStream = asyncStdoutStream;
        _asyncErrStream = asyncErrStream;
    }

    public CommandRunner(Command cmd)
        : this(cmd, false, NoInput, -1, DefaultTimeout, null, null)
    {
    }

    public CommandRunner(Command cmd, TimeSpan timeout)
        : this(cmd, false, NoInput, -1, timeout, null, null)
    {
    }

    /// <summary>Sets the verbose level for the command execution.</summary>
    public CommandRunner WithVerbose(bool verbose) =>
        new(_cmd, verbose, _input, _maxOutLogLines, _timeout, _asyncStdoutStream, _asyncErrStream);

    /// <summary>Sets the input for the command execution.</summary>
    public CommandRunner WithInput(byte[] input) =>
        new(_cmd, _verbose, input, _maxOutLogLines, _timeout, _asyncStdoutStream, _asyncErrStream);

    /// <summary>Sets the maximum number of output lines logged per stream.</summary>
    public CommandRunner WithMaxStdOutLogLines(int lines) =>
        new(_cmd, _verbose, _input, lines, _timeout, _asyncStdoutStream, _asyncErrStream);

    /// <summary>Sets a stream to redirect stdOut output to.</summary>
    public CommandRunner WithStdOutStream(Stream stream) =>
        new(_cmd, _verbose, _input, _maxOutLogLines, _timeout, stream, _asyncErrStream);

    /// <summary>Sets a stream to redirect stdErr output to.</summary>
    public CommandRunner WithStdErrStream(Stream stream) =>
        new(_cmd, _verbose, _input, _maxOutLogLines, _timeout, _asyncStdoutStream, stream);

    /// <summary>
    /// Executes the <see cref="Command"/> with the given input and writes to the console and the log
    /// depending on the exit code of the command and the verbose flag.
    /// </summary>
    public CommandOutputWithStatus Execute()
    {
        var stopwatch = Stopwatch.StartNew();
        var argv = _cmd.GetCommandLineElements();
        string startMsg = ShellUtils.PrettyPrintArgv(argv);
        startMsg = startMsg.Length > MaxCommandLength
            ? startMsg.Substring(0, MaxCommandLength) + "..."
            : startMsg;
        string validStartMsg = "Executing [" + startMsg + "]";
        Logger.LogInformation("{Msg}", validStartMsg);
        if (_verbose)
        {
            System.Console.Error.WriteLine(validStartMsg);
        }

        var stdoutCollector = new MemoryStream();
        var stderrCollector = new MemoryStream();
        if (_asyncStdoutStream != null)
        {
            byte[] note = Encoding.UTF8.GetBytes("stdOut redirected to external observer.");
            stdoutCollector.Write(note, 0, note.Length);
        }
        if (_asyncErrStream != null)
        {
            byte[] note = Encoding.UTF8.GetBytes("stdErr redirected to external observer.");
            stderrCollector.Write(note, 0, note.Length);
        }

        Stream stdoutStream = CommandOutputStream(_asyncStdoutStream ?? stdoutCollector, _maxOutLogLines);
        Stream stderrStream = CommandOutputStream(_asyncErrStream ?? stderrCollector, _maxOutLogLines);

        string commandName = argv[0];
        TerminationStatus? exitStatus = null;
        bool timedOut = false;
        try
        {
            var startInfo = new ProcessStartInfo
            {
                FileName = commandName,
                RedirectStandardInput = true,
                RedirectStandardOutput = true,
                RedirectStandardError = true,
                UseShellExecute = false,
                CreateNoWindow = true,
            };
            for (int i = 1; i < argv.Count; i++)
            {
                startInfo.ArgumentList.Add(argv[i]);
            }
            var env = _cmd.GetEnvironmentVariables();
            if (env != null)
            {
                startInfo.Environment.Clear();
                foreach (var kvp in env)
                {
                    startInfo.Environment[kvp.Key] = kvp.Value;
                }
            }
            var workingDir = _cmd.GetWorkingDirectory();
            if (workingDir != null)
            {
                startInfo.WorkingDirectory = workingDir;
            }

            using var process = new Process { StartInfo = startInfo };

            var stdoutDone = new ManualResetEventSlim(false);
            var stderrDone = new ManualResetEventSlim(false);

            try
            {
                process.Start();
            }
            catch (Exception e)
            {
                throw new CommandException(_cmd, e);
            }

            var pumpOut = PumpAsync(process.StandardOutput.BaseStream, stdoutStream, stdoutDone);
            var pumpErr = PumpAsync(process.StandardError.BaseStream, stderrStream, stderrDone);

            // Write input then close stdin.
            try
            {
                if (_input.Length > 0)
                {
                    process.StandardInput.BaseStream.Write(_input, 0, _input.Length);
                }
                process.StandardInput.BaseStream.Flush();
                process.StandardInput.Close();
            }
            catch (IOException)
            {
                // The process may have already exited and closed its stdin; ignore.
            }

            bool exited = process.WaitForExit((int)Math.Min(_timeout.TotalMilliseconds, int.MaxValue));
            if (!exited)
            {
                timedOut = true;
                try
                {
                    process.Kill(entireProcessTree: true);
                }
                catch (Exception)
                {
                    // best effort
                }
                process.WaitForExit();
            }

            // Ensure the output pumps drain fully after exit.
            stdoutDone.Wait();
            stderrDone.Wait();
            pumpOut.GetAwaiter().GetResult();
            pumpErr.GetAwaiter().GetResult();

            int rawResult = timedOut ? (128 + 15 /* SIGTERM */) : process.ExitCode;
            var status = new TerminationStatus(rawResult);
            exitStatus = status;

            byte[] stdoutBytes = stdoutCollector.ToArray();
            byte[] stderrBytes = stderrCollector.ToArray();

            if (timedOut)
            {
                var result = new CommandResult(stdoutBytes, stderrBytes, status);
                MaybeTreatTimeout(stdoutBytes, stderrBytes, true,
                    new AbnormalTerminationException(_cmd, result, status.ToString()));
            }

            if (!status.Success())
            {
                var result = new CommandResult(stdoutBytes, stderrBytes, status);
                string message =
                    $"Process '{commandName}' exited with status {status.ToShortString()}";
                throw new BadExitStatusWithOutputException(_cmd, result, message, stdoutBytes, stderrBytes);
            }

            return new CommandOutputWithStatus(status, stdoutBytes, stderrBytes);
        }
        finally
        {
            if (_maxOutLogLines != 0)
            {
                LogOutput(
                    LogLevel.Information, $"'{commandName}' STDOUT: ", stdoutCollector, _maxOutLogLines);
                LogOutput(
                    LogLevel.Information, $"'{commandName}' STDERR: ", stderrCollector, _maxOutLogLines);
            }

            string finishMsg;
            if (timedOut)
            {
                finishMsg = string.Format(
                    "Command '{0}' was killed after timeout. Execution time {1}. {2}",
                    commandName,
                    FormatDuration(stopwatch.Elapsed),
                    exitStatus != null ? exitStatus.ToString() : "(No exit status)");
                Logger.LogError("{Msg}", finishMsg);
            }
            else
            {
                finishMsg = string.Format(
                    "Command '{0}' finished in {1}. {2}",
                    commandName,
                    FormatDuration(stopwatch.Elapsed),
                    exitStatus != null ? exitStatus.ToString() : "(No exit status)");
                Logger.LogInformation("{Msg}", finishMsg);
            }
            if (_verbose)
            {
                System.Console.Error.WriteLine(finishMsg);
            }
        }
    }

    private static Task PumpAsync(Stream source, Stream destination, ManualResetEventSlim done)
    {
        return Task.Run(() =>
        {
            try
            {
                var buffer = new byte[8192];
                int read;
                while ((read = source.Read(buffer, 0, buffer.Length)) > 0)
                {
                    destination.Write(buffer, 0, read);
                }
                destination.Flush();
            }
            catch (Exception)
            {
                // Stream closed on process exit; ignore.
            }
            finally
            {
                done.Set();
            }
        });
    }

    /// <summary>
    /// Format a duration to a human-readable string. Assumes the duration is less than 24 hours,
    /// which should always be true for a command.
    /// </summary>
    private static string FormatDuration(TimeSpan duration) => duration.ToString(@"mm\:ss\.fff");

    private void MaybeTreatTimeout(
        byte[] stdout, byte[] stderr, bool hasTimedOut, AbnormalTerminationException e)
    {
        if (!hasTimedOut)
        {
            return;
        }
        string msg = string.Format(
            "Command '{0}' killed by Copybara after timeout ({1}s)."
                + " If this fails during a fetch or push use --repo-timeout flag."
                + " If this fails during checkout or another point, use --commands-timeout flag.\n"
                + "Exit info: {2}",
            _cmd.GetCommandLineElements()[0],
            (long)_timeout.TotalSeconds,
            e.GetResult().TerminationStatus);
        throw new CommandTimeoutException(
            e.GetCommand(), e.GetResult(), msg, stdout, stderr, _timeout);
    }

    /// <summary>Creates the OutputStream to feed the process output to.</summary>
    private Stream CommandOutputStream(Stream outputStream, int maxOutLogLines)
    {
        // If verbose we stream to the user console too.
        if (!_verbose)
        {
            return outputStream;
        }
        var console = System.Console.OpenStandardError();
        var limited = new LimitFilterOutputStream(
            console,
            // Assume a line has ~200 characters. If no limit, ~10k lines.
            maxOutLogLines > 0 ? maxOutLogLines * 200 : 10000 * 200,
            AfterLimitSuffix);
        return new MultiplexOutputStream(limited, outputStream);
    }

    private static void LogOutput(
        LogLevel level, string prefix, MemoryStream outputBytes, int maxLogLines)
    {
        string s = Encoding.UTF8.GetString(outputBytes.ToArray()).Trim();
        if (s.Length == 0)
        {
            return;
        }
        int lines = 0;
        foreach (var line in s.Split(Environment.NewLine))
        {
            Logger.Log(level, "{Prefix}{Line}", prefix, line);
            lines++;
            if (maxLogLines >= 0 && lines >= maxLogLines)
            {
                Logger.Log(level, "{Prefix}... truncated after {Max} line(s)", prefix, maxLogLines);
                break;
            }
        }
    }

    /// <summary>An <see cref="Stream"/> that writes to two underlying streams.</summary>
    private sealed class MultiplexOutputStream : Stream
    {
        private readonly Stream _s1;
        private readonly Stream _s2;

        public MultiplexOutputStream(Stream s1, Stream s2)
        {
            _s1 = s1;
            _s2 = s2;
        }

        public override bool CanRead => false;
        public override bool CanSeek => false;
        public override bool CanWrite => true;
        public override long Length => throw new NotSupportedException();

        public override long Position
        {
            get => throw new NotSupportedException();
            set => throw new NotSupportedException();
        }

        public override void Write(byte[] buffer, int offset, int count)
        {
            _s1.Write(buffer, offset, count);
            _s2.Write(buffer, offset, count);
        }

        public override void WriteByte(byte value)
        {
            _s1.WriteByte(value);
            _s2.WriteByte(value);
        }

        public override void Flush()
        {
            _s1.Flush();
            _s2.Flush();
        }

        public override int Read(byte[] buffer, int offset, int count) =>
            throw new NotSupportedException();

        public override long Seek(long offset, SeekOrigin origin) =>
            throw new NotSupportedException();

        public override void SetLength(long value) => throw new NotSupportedException();
    }
}
